src/server/replica.cc

```diff
@@ -515,25 +549,29 @@ error_code Replica::ConsumeRedisStream() {
 }
 
 error_code Replica::ConsumeDflyStream() {
-  // Send DFLY STARTSTABLE.
-  if (auto ec = SendNextPhaseRequest(); ec) {
-    return cntx_.Error(ec);
+  RETURN_ON_ERR(cntx_.Switch(absl::bind_front(&Replica::DefaultErrorHandler, this)));
+
+  // Transition flows into stable sync.
+  {
+    auto partition = Partition(num_df_flows_);
+    auto shard_cb = [&](unsigned index, auto*) {
+      const auto& local_ids = partition[index];
+      for (unsigned id : local_ids) {
+        auto ec = shard_flows_[id]->StartStableSyncFlow(&cntx_);
+        if (ec)
+          cntx_.Error(ec);
+      }
+    };
+
+    // Lock to prevent error handler from running on mixed state.
+    lock_guard lk{flows_op_mu_};
+    shard_set->pool()->AwaitFiberOnAll(std::move(shard_cb));
```

Comment:
In the cancllation flow if one flow sets error that this causes all flow to shut down without consuming the data that was already received in the socket right? is this the desired behaviour?

Reply:
we close the sockets anyway and retry the whole flow so I guess it should be fine

Reply:
It will not run until the guard holds the lock

Why should it not? Imagine a flow got an error, reported it, the handler run fast and cancelled the sockets, but not all flows actually created their sockets, so the "last ones to run" will not be affected by the cancellation at all. This can be solved by either moving initialization separately, checking for cancellation before inits or locking the initialization phase

Reply:
And yes, currently we just want to shut down everything and retry from scratch